File: app/src/main/java/com/averycorp/prismtask/domain/usecase/ModeBalanceTracker.kt
package com.averycorp.prismtask.domain.usecase

import com.averycorp.prismtask.data.local.entity.TaskEntity
import com.averycorp.prismtask.domain.model.TaskMode
import java.util.Calendar
import java.util.TimeZone

/**
 * Snapshot of the user's current mode balance — the Work / Play / Relax
 * version of [BalanceState], orthogonal to life-category balance.
 *
 * Numbers are descriptive, not prescriptive (see `docs/WORK_PLAY_RELAX.md`).
 */
data class ModeBalanceState(
    val currentRatios: Map<TaskMode, Float>,
    val rollingRatios: Map<TaskMode, Float>,
    val targetRatios: Map<TaskMode, Float>,
    val dominantMode: TaskMode,
    val totalTracked: Int
) {
    companion object {
        val EMPTY = ModeBalanceState(
            currentRatios = TaskMode.TRACKED.associateWith { 0f },
            rollingRatios = TaskMode.TRACKED.associateWith { 0f },
            targetRatios = TaskMode.TRACKED.associateWith { 1f / TaskMode.TRACKED.size },
            dominantMode = TaskMode.UNCATEGORIZED,
            totalTracked = 0
        )
    }
}

/**
 * Optional user-set targets for the mode balance bar. Defaults to an even
 * split. Targets do not drive overload notifications — the philosophy doc
 * is explicit that mode is descriptive-only — they only colour the
 * progress bars to indicate the user's chosen mix.
 */
data class ModeBalanceConfig(val workTarget: Float = 1f / 3f, val playTarget: Float = 1f / 3f, val relaxTarget: Float = 1f / 3f) {
    fun asMap(): Map<TaskMode, Float> = mapOf(
        TaskMode.WORK to workTarget,
        TaskMode.PLAY to playTarget,
        TaskMode.RELAX to relaxTarget
    )

    fun isValid(): Boolean {
        val sum = workTarget + playTarget + relaxTarget
        return kotlin.math.abs(sum - 1f) < 0.01f
    }
}

/**
 * Pure-function balance computation for the [TaskMode] dimension.
 *
 * Window cutoffs respect the user-configured Start-of-Day, mirroring the
 * SoD-aware cutoff in [BalanceTracker] (see PR #1060). Callers without
 * access to the SoD preference fall back to system midnight
 * (`dayStartHour = 0`).
 */
class ModeBalanceTracker {
    fun compute(
        allTasks: List<TaskEntity>,
        config: ModeBalanceConfig = ModeBalanceConfig(),
        now: Long = System.currentTimeMillis(),
        timeZone: TimeZone = TimeZone.getDefault(),
        dayStartHour: Int = 0,
        dayStartMinute: Int = 0
    ): ModeBalanceState {
        val weekCutoff = cutoff(now, days = 7, timeZone, dayStartHour, dayStartMinute)
        val monthCutoff = cutoff(now, days = 28, timeZone, dayStartHour, dayStartMinute)

        val current = computeRatios(allTasks, weekCutoff)
        val rolling = computeRatios(allTasks, monthCutoff)
        val total = countTracked(allTasks, weekCutoff)

        val dominant = if (total == 0) {
            TaskMode.UNCATEGORIZED
        } else {
            current.maxByOrNull { it.value }?.key ?: TaskMode.UNCATEGORIZED
        }

        return ModeBalanceState(
            currentRatios = current,
            rollingRatios = rolling,
            targetRatios = config.asMap(),
            dominantMode = dominant,
            totalTracked = total
        )
    }

    private fun computeRatios(tasks: List<TaskEntity>, cutoff: Long): Map<TaskMode, Float> {
        val counts = TaskMode.TRACKED.associateWith { 0 }.toMutableMap()
        var total = 0
        for (t in tasks) {
            val ts = timestampFor(t)
            if (ts < cutoff) continue
            val mode = TaskMode.fromStorage(t.taskMode)
            if (mode == TaskMode.UNCATEGORIZED) continue
            counts[mode] = (counts[mode] ?: 0) + 1
            total++
        }
        if (total == 0) return TaskMode.TRACKED.associateWith { 0f }
        return counts.mapValues { (_, count) -> count.toFloat() / total.toFloat() }
    }

    private fun countTracked(tasks: List<TaskEntity>, cutoff: Long): Int {
        var total = 0
        for (t in tasks) {
            val ts = timestampFor(t)
            if (ts < cutoff) continue
            val mode = TaskMode.fromStorage(t.taskMode)
            if (mode == TaskMode.UNCATEGORIZED) continue
            total++
        }
        return total
    }

    private fun timestampFor(task: TaskEntity): Long = task.completedAt ?: task.dueDate ?: task.createdAt

    private fun cutoff(
        now: Long,
        days: Int,
        timeZone: TimeZone,
        dayStartHour: Int,
        dayStartMinute: Int
    ): Long {
        val cal = Calendar.getInstance(timeZone)
        cal.timeInMillis = now
        val currentMinutesSinceMidnight = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE)
        val sodMinutesSinceMidnight = dayStartHour * 60 + dayStartMinute
        cal.set(Calendar.HOUR_OF_DAY, dayStartHour)
        cal.set(Calendar.MINUTE, dayStartMinute)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        if (currentMinutesSinceMidnight < sodMinutesSinceMidnight) {
            cal.add(Calendar.DAY_OF_YEAR, -1)
        }
        cal.add(Calendar.DAY_OF_YEAR, -(days - 1))
        return cal.timeInMillis
    }
}

File: web/src/utils/modeBalanceTracker.ts
import type { Task, TaskMode } from '@/types/task';

/**
 * Pure-function Task-Mode balance computations. Mirrors Android's
 * `ModeBalanceTracker.kt` (parity audit Pillar 3 — DEFERRED items
 * reclassified PROCEED 2026-05-15).
 *
 * Mode answers *what does the user want this task to produce?*
 *  - WORK produces output
 *  - PLAY produces enjoyment
 *  - RELAX produces restored energy
 *
 * Mode is orthogonal to `life_category` (the existing balanceTracker).
 * See `docs/WORK_PLAY_RELAX.md` for the philosophy doc, including the
 * descriptive-not-prescriptive copy rules and the SoD-aware window
 * semantics this tracker honors.
 */

/** Modes that participate in balance ratio computation. */
export const TRACKED_MODES: Exclude<TaskMode, 'UNCATEGORIZED'>[] = [
  'WORK',
  'PLAY',
  'RELAX',
];

export interface ModeBalanceState {
  /** Last 7 days of mode-tagged tasks, normalized 0..1. */
  currentRatios: Record<TaskMode, number>;
  /** Rolling 28-day average. */
  rollingRatios: Record<TaskMode, number>;
  /**
   * Configured targets per mode. Defaults to an even 1/3 split. Mode is
   * descriptive-only — targets do NOT drive overload notifications; they
   * only colour optional reference lines if a future UI surfaces them.
   */
  targetRatios: Record<TaskMode, number>;
  /**
   * Mode with the highest current ratio. `UNCATEGORIZED` when no data
   * (sentinel — never a "real" dominant value).
   */
  dominantMode: TaskMode;
  /** Number of tracked tasks contributing to currentRatios. */
  totalTracked: number;
}

/**
 * Optional user-set target ratios for the mode dimension. Defaults to an
 * even split across the three tracked modes. Mode is descriptive-only,
 * so targets are not user-tunable from Settings today — kept as a struct
 * to mirror Android's shape and to leave room for a future opt-in.
 */
export interface ModeBalanceConfig {
  workTarget: number;
  playTarget: number;
  relaxTarget: number;
}

export const DEFAULT_MODE_BALANCE_CONFIG: ModeBalanceConfig = {
  workTarget: 1 / 3,
  playTarget: 1 / 3,
  relaxTarget: 1 / 3,
};

function zeroRatios(): Record<TaskMode, number> {
  return {
    WORK: 0,
    PLAY: 0,
    RELAX: 0,
    UNCATEGORIZED: 0,
  };
}

export function modeConfigToMap(config: ModeBalanceConfig): Record<TaskMode, number> {
  return {
    WORK: config.workTarget,
    PLAY: config.playTarget,
    RELAX: config.relaxTarget,
    UNCATEGORIZED: 0,
  };
}

export const EMPTY_MODE_BALANCE_STATE: ModeBalanceState = {
  currentRatios: zeroRatios(),
  rollingRatios: zeroRatios(),
  targetRatios: modeConfigToMap(DEFAULT_MODE_BALANCE_CONFIG),
  dominantMode: 'UNCATEGORIZED',
  totalTracked: 0,
};

/** Returns true when the three mode targets sum to ~1.0 (within 0.01). */
export function isValidModeBalanceConfig(config: ModeBalanceConfig): boolean {
  const sum = config.workTarget + config.playTarget + config.relaxTarget;
  return Math.abs(sum - 1) < 0.01;
}

/**
 * Compute a {@link ModeBalanceState} from a pool of tasks.
 *
 * Window cutoffs respect the user-configured Start-of-Day so the "this
 * week" window matches the Today filter, habit streaks, and the
 * LifeCategory balance bar. Uncategorized tasks are excluded — they
 * never inflate or deflate a ratio.
 */
export function computeModeBalanceState(
  allTasks: Task[],
  config: ModeBalanceConfig = DEFAULT_MODE_BALANCE_CONFIG,
  options: { nowMs?: number; dayStartHour?: number; dayStartMinute?: number } = {},
): ModeBalanceState {
  const { nowMs = Date.now(), dayStartHour = 0, dayStartMinute = 0 } = options;
  const weekCutoff = cutoffMs(nowMs, 7, dayStartHour, dayStartMinute);
  const monthCutoff = cutoffMs(nowMs, 28, dayStartHour, dayStartMinute);

  const current = computeRatios(allTasks, weekCutoff);
  const rolling = computeRatios(allTasks, monthCutoff);
  const total = countTracked(allTasks, weekCutoff);

  let dominant: TaskMode = 'UNCATEGORIZED';
  if (total > 0) {
    let best = -1;
    for (const mode of TRACKED_MODES) {
      const r = current[mode] ?? 0;
      if (r > best) {
        best = r;
        dominant = mode;
      }
    }
  }

  return {
    currentRatios: current,
    rollingRatios: rolling,
    targetRatios: modeConfigToMap(config),
    dominantMode: dominant,
    totalTracked: total,
  };
}

function computeRatios(tasks: Task[], cutoff: number): Record<TaskMode, number> {
  const counts = zeroRatios();
  let total = 0;
  for (const t of tasks) {
    const ts = timestampFor(t);
    if (ts < cutoff) continue;
    const raw = (t.task_mode ?? 'UNCATEGORIZED') as TaskMode;
    if (raw === 'UNCATEGORIZED') continue;
    counts[raw] = (counts[raw] ?? 0) + 1;
    total++;
  }
  if (total === 0) return zeroRatios();
  for (const mode of TRACKED_MODES) {
    counts[mode] = (counts[mode] ?? 0) / total;
  }
  return counts;
}

function countTracked(tasks: Task[], cutoff: number): number {
  let total = 0;
  for (const t of tasks) {
    const ts = timestampFor(t);
    if (ts < cutoff) continue;
    const raw = (t.task_mode ?? 'UNCATEGORIZED') as TaskMode;
    if (raw === 'UNCATEGORIZED') continue;
    total++;
  }
  return total;
}

/**
 * Choose the most relevant timestamp for a task:
 *  - Completed tasks use completedAt.
 *  - Otherwise dueDate (parsed from ISO) if set, else createdAt.
 *
 * Mirrors `balanceTracker.timestampFor` so the mode + life-category
 * windows align day-for-day.
 */
function timestampFor(task: Task): number {
  const completed = toMs(task.completed_at);
  if (completed != null) return completed;
  const due = toMs(task.due_date);
  if (due != null) return due;
  const created = toMs(task.created_at);
  return created ?? 0;
}

function toMs(v: string | number | null | undefined): number | null {
  if (v == null) return null;
  if (typeof v === 'number') return v;
  const parsed = Date.parse(v);
  return Number.isNaN(parsed) ? null : parsed;
}

/**
 * Lower bound of the balance window. Snaps `now` back to the most recent
 * day-start, then walks back `days - 1` days so the window covers `days`
 * logical days inclusive of today.
 */
function cutoffMs(
  now: number,
  days: number,
  dayStartHour: number,
  dayStartMinute: number,
): number {
  const d = new Date(now);
  const minutesNow = d.getHours() * 60 + d.getMinutes();
  const sodMinutes = dayStartHour * 60 + dayStartMinute;
  d.setHours(dayStartHour, dayStartMinute, 0, 0);
  if (minutesNow < sodMinutes) {
    d.setDate(d.getDate() - 1);
  }
  d.setDate(d.getDate() - (days - 1));
  return d.getTime();
}
